### PR TITLE
request to axios

### PR DESCRIPTION
**Pull request by soudher1**
Ref# https://github.com/github/npm/issues/12534

This pull request primarily focuses on transitioning from the `request` library to `axios` for making HTTP requests in the `index.js` file of the project. This change impacts various functions in the file, including how HTTP requests are made and how responses are handled. Additionally, the project's version is incremented in the `package.json` file.

Major changes:

* [`index.js`](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L3-R4): Replaced `request` with `axios` for making HTTP requests. This change required updates to the `Client.prototype._resolveToken` and `Client.prototype._request2` methods to handle the differences between the two libraries. [[1]](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L3-R4) [[2]](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R331-R337) [[3]](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R349-R379) [[4]](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R390-R403)
* [`index.js`](diffhunk://#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L44-R44): Updated the regular expression in the `Client` function to escape the period character.
* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L3-R8): Updated the project version from 2.4.5 to 2.4.6 and replaced the `request` dependency with `axios`.

### CODE REVIEW

**Comment by soudher1** [comment]
.lock --global

File: index.js [hunk 1]
```diff
@@ -41,7 +41,7 @@ var Client = function (host, opts) {
 
   if (!opts) opts = {}
 
-  var disc = opts.token || /^https:\/\/discovery.etcd.io\//.test(host || '')
+  var disc = opts.token || /^https:\/\/discovery\.etcd\.io\//.test(host || '')
```
Comment (wraithgar) [inline index.js]:
```suggestion
  var disc = opts.token || host?.startsWith('https://discover.etcd.io/')
```

File: index.js [hunk 2]
```diff
@@ -345,54 +346,61 @@ Client.prototype._resolveToken = function (cb) {
 
     done()
   })
+  .catch(err => {return done(err)})
```
Comment (wraithgar) [inline index.js:353]: What errors is this supposed to be catching?

Comment (soudher1) [reply index.js:353]: 3 types of errors are caught:
1. The request was made and the server responded with a status code that falls out of the range of 2xx (error.response holds information)
2. The request was made but no response was received (error.request holds the information)
3. Something happened in setting up the request that triggered an Error (error.message could give more information)

Comment (wraithgar) [reply index.js:353]: All of those are something axios would handle internally though right? Unless I am mistaken the abort controller is intended to send an abort signal into axios when things outside of the request itself go wrong and we need to abort the request.

Comment (soudher1) [reply index.js:353]: I think abort is slightly different solution where user chooses to cancel the request.
In this case, we get a handle for cancelling the in-progress request. If we cancel the request, say for a timeout scenario, some debug information is not obtained if we skip the catch block.

**Review by wraithgar — APPROVED** [review]

File: index.js (hunk 1 above)
Comment (shmam) [reply index.js]: turns out this introduces a breaking change when host is an array (this came up in tests with furthermore). Can we use .test or is there an alternative that allows us to check if host is an array ?

Comment (wraithgar) [reply index.js]: Yikes.  It was working on accident before.  `.test` explicitly takes a string, and was casting the array to a string under the hood.

It'd be best if we had code above this to make sure `opts` is the exact type we want.  Typically that's done by checking if it's a string, and then moving it into an array.

Then you can use `.some` with `.startsWith` e.g.

```javascript
if (typeof host === 'string') {
  host = [host]
}
host.some(h => h.startsWith('foo'))
```

Comment (shmam) [reply index.js]: perf! thanks for the suggestion with greater context on what it **should** be doing

File: index.js [hunk 3]
```diff
@@ -41,7 +41,9 @@ var Client = function (host, opts) {
 
   if (!opts) opts = {}
 
-  var disc = opts.token || /^https:\/\/discovery\.etcd\.io\//.test(host || '');
+  if (typeof host === 'string') host = [host]
+
+  var disc = opts.token || host?.some(h => h.startsWith('http://discovery.etcd.io'))
```
Comment (wraithgar) [inline index.js]:
```suggestion
  var disc = opts.token || host?.some(h => h.startsWith('http://discovery.etcd.io/'))
```

Comment (wraithgar) [reply index.js]: Thanks codeql!

**Pull request merged**